pkg/agent/proxy/proxier.go

```diff
@@ -130,19 +136,46 @@ func getBindingProtoForIPProto(endpointIP string, protocol corev1.Protocol) bind
 	return bindingProtocol
 }
 
+// removeEndpoint removes flows for the given Endpoint from the data path if these flows are no longer
+// needed by any Service. Endpoints from different Services can have the same characteristics and thus
+// can share the same flows. removeEndpoint must be called whenever an Endpoint is no longer used by a
+// given Service. If the Endpoint is still referenced by any other Services, no flow will be removed.
+// The method only returns an error if a data path operation fails. If the flows are successfully
+// removed from the data path, the method returns true. Otherwise, if the flows are still needed for
+// other Services, it returns false.
+func (p *proxier) removeEndpoint(endpoint k8sproxy.Endpoint, protocol binding.Protocol) (bool, error) {
+	key := endpointKey(endpoint, protocol)
+	count := p.endpointReferenceCounter[key]
+	if count == 1 {
+		if err := p.ofClient.UninstallEndpointFlows(protocol, endpoint); err != nil {
+			return false, err
+		}
+		delete(p.endpointReferenceCounter, key)
+	} else if count > 1 {
+		p.endpointReferenceCounter[key] = count - 1
+		return false, nil
+	}
+	return true, nil
+}
+
 func (p *proxier) removeStaleEndpoints(staleEndpoints map[k8sproxy.ServicePortName]map[string]k8sproxy.Endpoint) {
 	for svcPortName, endpoints := range staleEndpoints {
 		for _, endpoint := range endpoints {
-			bindingProtocol := getBindingProtoForIPProto(endpoint.IP(), svcPortName.Protocol)
-			if err := p.ofClient.UninstallEndpointFlows(bindingProtocol, endpoint); err != nil {
+			klog.V(2).Infof("Removing stale Endpoint %s of Service %s", endpoint.String(), svcPortName.String())
+			removed, err := p.removeEndpoint(endpoint, getBindingProtoForIPProto(endpoint.IP(), svcPortName.Protocol))
+			if err != nil {
 				klog.Errorf("Error when removing Endpoint %v for %v", endpoint, svcPortName)
 				continue
 			}
-			if m, ok := p.endpointInstalledMap[svcPortName]; ok {
+			if removed {
+				m := p.endpointInstalledMap[svcPortName]
```

Comment:
If we don't remove the endpoint from this servicePortName's map when it's referenced by other services, this `m` will never get to 0 and will be leaked finally?

Reply:
Thanks for mentioning. Updated to a new function.